Source/Core/Common/FileSearch.cpp

```diff
@@ -124,20 +63,34 @@ std::vector<std::string> DoFileSearch(const std::vector<std::string>& directorie
   };
   for (const auto& directory : directories)
   {
-    fs::path directory_path = StringToPath(directory);
-    if (fs::is_directory(directory_path))  // Can't create iterators for non-existant directories
+#ifdef ANDROID
+    // Having a specialized implementation for Android content provides a much needed performance
+    // boost.
```

Comment:
This is no longer just a performance boost, it's required in order for the game list to work at all on Android. The code is fine, but please change the comment.

Reply:
I'm not 100% sure what you're asking with that last sentence, but let me try to go through what happens during a typical call to `ScanDirectoryTree` on Android.

Let's say that the user has selected a directory to use for the game list, and let's say we're going to search through it using `ScanDirectoryTree`. (In practice it would actually use `DoFileSearchAndroidContent` through `DoFileSearch`, but for the sake of example, we'll ignore that.) Android shows the user a SAF file picker, and once that file picker delivers a result to the game list activity, we're given a `Uri` object that looks like this (actual example from my phone): `content://com.android.externalstorage.documents/tree/primary%3AGames%2FDolphin`

(That last component is actually a URI-encoded path on external storage, but we're not allowed to rely on it being structured anything like a path. It's different for content providers other than `com.android.externalstorage.documents` – for instance, I recall Google Drive just putting a bunch of hex digits or something there.)

Now we're going to call `ScanDirectoryTree`, and here's where it gets real janky. The input path is `content://com.android.externalstorage.documents/tree/primary%3AGames%2FDolphin`. `ScanDirectoryTree` calls `IsPathAndroidContent` and gets a result of true, making it call `GetAndroidContentChildNames`. Let's say the directory contains a file named `Super Paper Mario.iso` and a folder called `GameCube`. `GetAndroidContentChildNames` will return `Super Paper Mario.iso` and `GameCube`. `ScanDirectoryTree` computes `physical_name` as `directory + DIR_SEP + virtual_name`, so `physical_name` for our file and folder will be `content://com.android.externalstorage.documents/tree/primary%3AGames%2FDolphin/Super Paper Mario.iso` and `content://com.android.externalstorage.documents/tree/primary%3AGames%2FDolphin/GameCube` respectively. These are not actually valid content URIs! But once we send these "mangled" URIs to Java land, Dolphin's code will unmangle them into real content URIs that Android knows how to deal with. So for instance, if `ScanDirectoryTree` recursively calls `ScanDirectoryTree` for `content://com.android.externalstorage.documents/tree/primary%3AGames%2FDolphin/GameCube`, the new call to `DoFileSearchAndroidContent` will internally do unmangling (so that it can request the contents of the directory. The same goes for when it's time to actually run the game – `IOFile` will call Java code to open the file, and that Java code unmangles the mangled URI we passed it.

We could make this less janky and get rid of the mangling/unmangling if we made a wrapper class for paths and made every bit of code in Dolphin use it, but I'm honestly not sure if that's better or worse.

Now you know one of the reasons why I like to complain about Android :)

Reply:
Yeah, when we're doing recursive calls we only really have to check `IsPathAndroidContent` on the top-level call. It'll always return the same results for the recursive calls as for the top-level call.